### PR TITLE
Keystore implementation for substrate chain

### PR DESCRIPTION
**Pull request by oshorefueled**
This PR implements a keystore for the substrate chain and the keystore provider. It also stubs out the provider methods which would be implemented in subsequent PRs. Currently, they all implement panic statements.

### CODE REVIEW

File: relayer/chains/substrate/key.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	keybase, err := keystore.New(sp.PCfg.ChainID, sp.PCfg.KeyringBackend, sp.PCfg.KeyDirectory, sp.Input)
+	if err != nil {
+		return err
+	}
+	sp.Keybase = keybase
+	return nil
+}
+
+func (sp *SubstrateProvider) KeystoreCreated(path string) bool {
+	if _, err := os.Stat(sp.PCfg.KeyDirectory); errors.Is(err, os.ErrNotExist) {
```
Comment (jtieri) [inline relayer/chains/substrate/key.go]: We can drop the `errors.Is` check along with the `else` branch here since we are returning `false` no matter what the error type is.

File: relayer/chains/substrate/key.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+func (sp *SubstrateProvider) KeyExists(name string) bool {
+	if sp.Keybase == nil {
+		return false
+	}
+
+	k, err := sp.Keybase.Key(name)
+	if err != nil {
+		fmt.Println(err)
```
Comment (jtieri) [inline relayer/chains/substrate/key.go]: We can remove this `Println` call. Arguably this function may be better off returning an error vs a bool but I think that is outside the scope of this PR.

File: relayer/chains/substrate/key.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+		return false
+	}
+	return k.GetName() == name
+}
+
+func (sp *SubstrateProvider) ExportPrivKeyArmor(keyName string) (armor string, err error) {
+	// TODO
+	panic("implement me -> ExportPrivKeyArmor -> https://github.com/ComposableFi/relayer/issues/6")
```
Comment (jtieri) [inline relayer/chains/substrate/key.go]: Ahh yes, I forgot we never circled back up regarding this.

@jackzampolin is this `ExportPrivKeyArmor` function a necessary function in the `KeyProvider`? We were all a bit confused on the exact purpose of this function or when it would be useful as a user.

https://github.com/ComposableFi/relayer/issues/6

File: relayer/chains/substrate/keystore/errors.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package keystore
+
+const (
+	ErrTextKeyNotFound             = "key not found: %v"
+	ErrTextKeyWithAddressNotFound  = "key with address %s not found"
+	ErrTextAddressExists           = "account with address %s already exists in keyring, delete the key first if you want to recreate it"
+	ErrTextPubkeyExists            = "public key already exists in keybase"
+	ErrTextUnknownKeyringBackend   = "unknown keyring backend %v"
+	ErrTextFailedToRead            = "failed to read %s: %v"
+	ErrTextFailedToOpen            = "failed to open %s: %v"
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/errors.go]: While it's not invalid to add the formatting verbs directly into the error strings like this, it does create a situation where developers need to know which format verbs are included before hand in order for the formatting to turn out correctly.

I think the better approach here is to use plain strings and word them in a way where one could easily wrap the strings when using them 

e.g. 
```
const ErrTextKeyNotFound = "key not found"

fmt.Errorf("%s: %s", ErrTextKeyNotFound, keyName)
```

File: relayer/chains/substrate/keystore/types.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+type keystore struct {
+	db keyring.Keyring
+}
+
+// Keyring exposes operations over a backend supported by github.com/99designs/keyring.
+type Keyring interface {
+	// List all keys.
+	List() ([]Info, error)
+
+	// Key and KeyByAddress return keys by uid and address respectively.
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/types.go]: Is this comment still applicable? I don't see a `KeyByAddress` function in the interface anymore.

File: relayer/chains/substrate/keystore/types.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	Key(uid string) (Info, error)
+
+	// Delete and DeleteByAddress remove keys from the keyring.
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/types.go]: Is this comment still applicable? I don't see a `DeleteByAddress` function in the interface anymore.

File: relayer/chains/substrate/keystore/types.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	Delete(uid string) error
+
+	// NewAccount converts a mnemonic to a private key and BIP-39 HD Path and persists it.
+	// It fails if there is an existing key Info with the same address.
+	NewAccount(name, mnemonic string, network uint8) (Info, error)
+}
+
+// Info is the publicly exposed information about a keypair
+type Info interface {
+	// Name of the key
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/types.go]: Proper godoc style comments begin with the function/variable name, so these should all be adjusted to begin with the function names. e.g. `GetName returns the name of the underlying key...`

File: relayer/chains/substrate/keystore/types.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	// Address
+	GetAddress() string
+	// Public key
+	GetPublicKey() []byte
+	// KeyPair
+	GetKeyringPair() signature.KeyringPair
+}
+
+// localInfo is the public information about a locally stored key
+// Note: Algo must be last field in struct for backwards amino compatibility
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/types.go]: Should this comment be updated? I don't see an `Algo` field in the struct.

File: relayer/chains/substrate/keystore/info.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
+package keystore
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ComposableFi/go-substrate-rpc-client/v4/signature"
+)
+
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/info.go:9]: Since the methods for these receivers are written in this file vs. the `types.go` file where the types actually live, I think it makes more sense to bring those types into this file also so the types live in the same place as the methods on them.

File: relayer/chains/substrate/keystore/info.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+func newLocalInfo(name string, kp signature.KeyringPair) (Info, error) {
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/info.go]: Does this function signature need to include an `error` in the return types since there is never a case where an error is returned?

File: relayer/chains/substrate/keystore/info.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+	}, nil
+
+}
+
+// GetType implements Info interface
+func (i localInfo) GetAddress() string {
+	return i.Keypair.Address
+}
+
+// GetType implements Info interface
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/info.go]: Duplicate comment here, same as line 19 just above this.

File: relayer/chains/substrate/provider_test.go [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+const homePath = "/tmp"
+
+func getSubstrateConfig(keyDir string, network uint8) *substrate.SubstrateProviderConfig {
+	return &substrate.SubstrateProviderConfig{
+		Key:            "default",
+		ChainID:        "substrate-test",
+		KeyringBackend: keystore.BackendTest,
+		KeyDirectory:   keyDir,
+		Timeout:        "20s",
+		Network:        network,
```
Comment (jtieri) [inline relayer/chains/substrate/provider_test.go:17]: What does the `Network` field represent?

File: relayer/chains/substrate/provider.go [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+}
+
+func (spc SubstrateProviderConfig) Validate() error {
+	if _, err := time.ParseDuration(spc.Timeout); err != nil {
+		return fmt.Errorf("invalid Timeout: %w", err)
+	}
+	return nil
+}
+
+func keysDir(home, chainID string) string {
```
Comment (jtieri) [inline relayer/chains/substrate/provider.go:49]: Since this function is in use to derive the keys directory I think we could drop the `key-directory` field from the `SubstrateProviderConfig`

File: relayer/chains/substrate/provider.go [hunk 14]
```diff
@@ -0,0 +1,468 @@
...
+	_ provider.KeyProvider    = &SubstrateProvider{}
+	_ provider.ProviderConfig = &SubstrateProviderConfig{}
+)
+
+type SubstrateProviderConfig struct {
+	Key            string `json:"key" yaml:"key"`
+	ChainName      string `json:"-" yaml:"-"`
+	ChainID        string `json:"chain-id" yaml:"chain-id"`
+	RPCAddr        string `json:"rpc-addr" yaml:"rpc-addr"`
+	RelayRPCAddr   string `json:"relay-rpc-addr" yaml:"rpc-addr"`
```
Comment (jtieri) [inline relayer/chains/substrate/provider.go]: There seems to be some subtlety between these two fields that is not immediately clear to me. What is the difference between their usage?

File: relayer/chains/substrate/provider.go [hunk 15]
```diff
@@ -0,0 +1,468 @@
...
+	AccountPrefix  string `json:"account-prefix" yaml:"account-prefix"`
+	KeyringBackend string `json:"keyring-backend" yaml:"keyring-backend"`
+	KeyDirectory   string `json:"key-directory" yaml:"key-directory"`
+	Debug          bool   `json:"debug" yaml:"debug"`
+	Timeout        string `json:"timeout" yaml:"timeout"`
+	Network        uint8  `json:"network" yaml:"network"`
+}
+
+func (spc SubstrateProviderConfig) Validate() error {
```
Comment (jtieri) [inline relayer/chains/substrate/provider.go:42]: It doesn't look like `Validate` gets called when initializing a new `SubstrateProvider`. Perhaps this happens elsewhere or in a subsequent PR but I figured I would point this out.

File: relayer/chains/substrate/provider.go [hunk 16]
```diff
@@ -0,0 +1,468 @@
...
+	if err != nil {
+		return err
+	}
+
+	sp.Keybase = keybase
+	return nil
+}
+
+type SubstrateProvider struct {
+	log     *zap.Logger
```
Comment (jtieri) [inline relayer/chains/substrate/provider.go:87]: It looks like the logger never gets initialized when the `SubstrateProvider` is initialized.
reference to where we do this in the `CosmosProvider`, https://github.com/cosmos/relayer/blob/d59c98adb368822a25328fe74db433f8aae78a1d/relayer/chains/cosmos/provider.go#L54

File: relayer/chains/substrate/keystore/keyring.go [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+				continue
+			}
+
+			passwordHash, err := bcrypt.GenerateFromPassword([]byte(pass), 2)
+			if err != nil {
+				fmt.Fprintln(os.Stderr, err)
+				continue
+			}
+
+			if err := ioutil.WriteFile(dir+"/keyhash", passwordHash, 0555); err != nil {
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/keyring.go]: `filepath.Join` is a better choice here instead of `dir+"/keyhash"`

File: relayer/chains/substrate/keystore/keyring.go [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = ks.db.Remove(info.GetAddress())
+	if err != nil {
+		return err
+	}
+
+	err = ks.db.Remove(infoKey(name))
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/keyring.go:223]: There is probably a good reason for another call to `Remove` here since you seem to be passing in a new argument but I'm not exactly sure what's going on by just following along so figured I would ask for some clarification.

File: relayer/chains/substrate/keystore/keyring.go [hunk 19]
```diff
@@ -0,0 +1,338 @@
...
+
+	keys, err := ks.db.Keys()
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Strings(keys)
+
+	for _, key := range keys {
+		if strings.HasSuffix(key, infoSuffix) {
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/keyring.go]: This check could be inverted to remove some overhead from reading the code inside the block.

i.e. 
```
if !strings.HasSuffix(key, infoSuffix) {
    continue
}

(rest of the code that was previously inside of the if statement)
```

File: relayer/chains/substrate/keystore/keyring.go [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+
+func (ks keystore) Key(uid string) (Info, error) {
+	infoKey := infoKey(uid)
+	return ks.key(infoKey)
+}
+
+func (ks keystore) NewAccount(name string, mnemonic string, network uint8) (Info, error) {
+	kp, err := signature.KeyringPairFromSecret(mnemonic, network)
+	if err != nil {
+		fmt.Println("error while creating keypair: " + err.Error())
```
Comment (jtieri) [inline relayer/chains/substrate/keystore/keyring.go]: I believe this should be `return nil, fmt.Errorf("failed to create keypair: %w", err)`

**Review by jtieri — COMMENTED** [review]
Good work @oshorefueled and the rest of the Composable team! Feels like a historic milestone getting this merged.

I've got a handful of nits and some questions here and there but this is shaping up nicely.

One thing I want to point out is that the tests are currently failing because of an issue with the `go.mod` file. Due to the presence of the `_test` directory, that is home to the integration style tests, there is a bit of nuance to running `go mod tidy` in the relayer. Currently you need to navigate to the root directory of the relayer and run `mv _test test && go mod tidy && mv test _test`, in my experience this needs to be run straight from a terminal window *not* inside of an IDE because the IDE does some behind the scenes magic that prevents the command from running properly. This directory will be removed as soon as we have parity with the tests in this directory and `ibctest`, which will hopefully be in the next week or so seeing as we only need to replicate one more test case in `ibctest`.

tl;dr: To fix the failing tests you need to run the following commands from a terminal window that is outside of an IDE.

```
cd path/to/relayer/root/dir
mv _test test && go mod tidy && mv test _test
```

File: relayer/chains/substrate/keystore/keyring.go (hunk 18 above)
Comment (oshorefueled) [reply relayer/chains/substrate/keystore/keyring.go:223]: creating a new account in the keystore saves it to storage by name and address. That way, there won't be a duplicate keypair if a user tried to import the same keypair with different key names. When deleting that keypair it also needs to be deleted by name and address, hence the repeated remove statements.

File: relayer/chains/substrate/provider_test.go (hunk 12 above)
Comment (oshorefueled) [reply relayer/chains/substrate/provider_test.go:17]: Network specifies the address format to be used when generating addresses from a public key. So different networks in substrate can use the same public keys but they usually have different address formats. https://docs.substrate.io/reference/command-line-tools/subkey/#:~:text=for%20keystore%20services-,%2Dn%2C%20%2D%2Dnetwork%20%3Cnetwork%3E,-Specifies%20the%20network

File: relayer/chains/substrate/provider.go (hunk 14 above)
Comment (oshorefueled) [reply relayer/chains/substrate/provider.go]: RPCAddr is the address of the composable parachain while RelayRPCAddr is the address of the relay chain.

File: relayer/chains/substrate/provider.go (hunk 13 above)
Comment (oshorefueled) [reply relayer/chains/substrate/provider.go:49]: We could do that if the home directory field is part of the provider struct. The keysDir uses the home path to derive the directory for the keys. The key-directory is used in the Init method to create a new keystore https://github.com/ComposableFi/relayer/blob/sub_keystore/relayer/chains/substrate/provider.go#L70. However, that field isn't accessible from the Init function except through the config.

File: relayer/chains/substrate/provider.go (hunk 15 above)
Comment (oshorefueled) [reply relayer/chains/substrate/provider.go:42]: yeaah, didn't think it was important for the keystore implementation. We could add it now though, no biggie.

File: relayer/chains/substrate/provider.go (hunk 16 above)
Comment (oshorefueled) [reply relayer/chains/substrate/provider.go:87]: I think a new log is initialized for the cosmos provider because it uses the ChainClient implementation from lens. We should be able to just pass the log to the SubstrateProvider like here right? https://github.com/cosmos/relayer/blob/d59c98adb368822a25328fe74db433f8aae78a1d/relayer/chains/cosmos/provider.go#L65
